### ISSUE TITLE
BYOL implementation, loss function seems to be wrong

### ISSUE THREAD

**Issue by FrancescoSaverioZuppichini** [PyTorchLightning/lightning-bolts#697]
Hi guys, 

I hope you are doing great. In the BYOL implementation [here](https://github.com/PyTorchLightning/lightning-bolts/blob/e22b8d0efbfd0a15194446170b193ec450443666/pl_bolts/models/self_supervised/byol/byol_module.py#L118)
you are not creating the loss based on the similarity between `z1` and `z2` but between `h1` and `z2` and vice-versa down the line. I think this is wrong

```python
def shared_step(self, batch, batch_idx):
        imgs, y = batch
        img_1, img_2 = imgs[:2]

        # Image 1 to image 2 loss
        y1, z1, h1 = self.online_network(img_1)
        with torch.no_grad():
            y2, z2, h2 = self.target_network(img_2)
        loss_a = -2 * F.cosine_similarity(h1, z2).mean()

        # Image 2 to image 1 loss
        y1, z1, h1 = self.online_network(img_2)
        with torch.no_grad():
            y2, z2, h2 = self.target_network(img_1)
        # L2 normalize
        loss_b = -2 * F.cosine_similarity(h1, z2).mean()

        # Final loss
        total_loss = loss_a + loss_b

        return loss_a, loss_b, total_loss
```

Thank you.

Francesco

**Comment by FrancescoSaverioZuppichini** [comment]
sorry guys, it is correct :)

**Comment by MaveriQ** [comment]
@FrancescoSaverioZuppichini On a related note, shouldn't there be a detach() with z2 to stop gradient flow in target network?

**Comment by Borda** [comment]
@ananyahjha93 ^^
